### PR TITLE
`Development`: Improve JPQL style for several repositories

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

### Motivation and Context

I just fixed the style in one repository so I decided to go through some more repos to make the codebase more consistent and clean.

### Description

I enforced capitalisation (SELECT instead of select), simple variable declaration (:var instead of :#{#var}, explicit enum values, consistent newlines and indentations

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2

### Test Coverage

No functional changes.

## Summary by CodeRabbit

- **New Features**
	- Enhanced user retrieval across the system to consider group memberships and exclude deleted users, improving accuracy and performance.
- **Refactor**
	- Updated queries in various repositories for better readability, maintainability, and performance.
	- Modified service methods to use updated repository methods, ensuring more precise user selection based on group membership and deletion status.
- **Bug Fixes**
	- Fixed issues related to incorrect user counts in courses by excluding deleted users.
- **Tests**
	- Adjusted test cases to align with the new user retrieval logic, ensuring accurate testing scenarios.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 1]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
-            order by r.completionDate
+            SELECT new de.tum.in.www1.artemis.domain.statistics.StatisticsEntry(
+                r.completionDate, COUNT(r.id)
+            )
+            FROM Result r
+            WHERE r.completionDate >= :startDate
+                AND r.completionDate <= :endDate
+                AND r.participation.exercise.id in :exerciseIds
+            GROUP BY r.completionDate
+            ORDER BY r.completionDate
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:384]: While we want to capitalized SQL keywords (e.g. SELECT, FROM, WHERE, JOIN), we don't want to capitalize table names (e.g. Submission, Result) or the new keyword.

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 2]
```diff
@@ -29,67 +29,76 @@
...
+            )
+            FROM Submission s
+            WHERE s.submissionDate >= :startDate
+                AND s.submissionDate <= :endDate
+                AND s.participation.exercise.id = :exerciseId
+            GROUP BY s.submissionDate
+            ORDER BY s.submissionDate ASC
             """)
     List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate,
             @Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:76]:
```suggestion
            @Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 3]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+            )
+            FROM StudentParticipation p
+                LEFT JOIN p.submissions submission
+            WHERE submission.submissionDate >= :startDate
+                AND submission.submissionDate <= :endDate
+                AND p.student.login NOT LIKE '%test%'
+            AND p.exercise.id = :exerciseId
+            ORDER BY submission.submissionDate ASC
             """)
     List<StatisticsEntry> getActiveUsersForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:156]:
```suggestion
    List<StatisticsEntry> getActiveUsersForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 4]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+                e.endDate, COUNT(e.id)
+            )
+            FROM Exam e
+            WHERE e.endDate >= :startDate
+                AND e.endDate <= :endDate
+                AND e.course.id = :courseId
+            GROUP BY e.endDate
+            ORDER BY e.endDate ASC
             """)
     List<StatisticsEntry> getConductedExamsForCourse(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") Long courseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:249]:
```suggestion
    List<StatisticsEntry> getConductedExamsForCourse(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") long courseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 5]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+            FROM Exam e
+                LEFT JOIN e.studentExams se
+            WHERE e.endDate >= :startDate
+                AND e.endDate <= :endDate
+                AND se.submitted IS TRUE
+                AND e.course.id = :courseId
+            GROUP BY e.endDate
+            ORDER BY e.endDate ASC
             """)
     List<StatisticsEntry> getExamParticipationsForCourse(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") Long courseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:279]:
```suggestion
    List<StatisticsEntry> getExamParticipationsForCourse(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") long courseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 6]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+                e.endDate, SUM(SIZE(e.examUsers))
+            )
+            FROM Exam e
+            WHERE e.endDate >= :startDate
+                AND e.endDate <= :endDate
+                AND e.course.id = :courseId
+            GROUP BY e.endDate
+            ORDER BY e.endDate ASC
             """)
     List<StatisticsEntry> getExamRegistrationsForCourse(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") Long courseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:305]:
```suggestion
    List<StatisticsEntry> getExamRegistrationsForCourse(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") long courseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 7]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+            FROM Result r
+            WHERE r.completionDate >= :startDate
+                AND r.completionDate <= :endDate
+                AND (
+                    r.assessmentType = de.tum.in.www1.artemis.domain.enumeration.AssessmentType.MANUAL
+                    OR r.assessmentType = de.tum.in.www1.artemis.domain.enumeration.AssessmentType.SEMI_AUTOMATIC
+                ) AND r.assessor.login NOT LIKE '%test%'
+                AND r.participation.exercise.id = :exerciseId
             """)
     List<StatisticsEntry> getActiveTutorsForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:357]:
```suggestion
    List<StatisticsEntry> getActiveTutorsForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 8]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+                r.completionDate, COUNT(r.id)
+            )
+            FROM Result r
+            WHERE r.completionDate >= :startDate
+                AND r.completionDate <= :endDate
+                AND r.participation.exercise.id = :exerciseId
+            GROUP BY r.completionDate
+            ORDER BY r.completionDate
             """)
     List<StatisticsEntry> getCreatedResultsForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:400]:
```suggestion
    List<StatisticsEntry> getCreatedResultsForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 9]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+                r.completionDate, SUM(SIZE(r.feedbacks))
+            )
+            FROM Result r
+            WHERE r.completionDate >= :startDate
+                AND r.completionDate <= :endDate
+                AND r.participation.exercise.id = :exerciseId
+            GROUP BY r.completionDate
+            ORDER BY r.completionDate
             """)
     List<StatisticsEntry> getResultFeedbacksForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:442]:
```suggestion
    List<StatisticsEntry> getResultFeedbacksForExercise(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 10]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+            FROM Post post
+                LEFT JOIN TREAT (post.conversation AS Channel) channel
+            WHERE post.creationDate >= :startDate
+                AND post.creationDate <= :endDate
+                AND channel.course.id = :courseId
+                AND channel.isCourseWide IS TRUE
+            GROUP BY post.creationDate
+            ORDER BY post.creationDate ASC
             """)
     List<StatisticsEntry> getPostsForCourseInDateRange(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") Long courseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:457]:
```suggestion
    List<StatisticsEntry> getPostsForCourseInDateRange(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("courseId") long courseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 11]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
+            FROM Post post
+                LEFT JOIN TREAT (post.conversation AS Channel) channel
+            WHERE post.creationDate >= :startDate
+                AND post.creationDate <= :endDate
+                AND channel.exercise.id = :exerciseId
+            GROUP BY post.creationDate
+            ORDER BY post.creationDate ASC
             """)
     List<StatisticsEntry> getPostsForExerciseInDateRange(@Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate,
             @Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:472]:
```suggestion
            @Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 12]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
     @Query("""
-            select count(post)
-            from Post post left join Channel channel ON channel.id = post.conversation.id
-            where channel.exercise.id = :#{#exerciseId}
+            SELECT COUNT(post)
+            FROM Post post
+                LEFT JOIN TREAT (post.conversation AS Channel) channel
+            WHERE channel.exercise.id = :exerciseId
             """)
     long getNumberOfExercisePosts(@Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:480]:
```suggestion
    long getNumberOfExercisePosts(@Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java [hunk 13]
```diff
@@ -102,374 +111,442 @@ List<StatisticsEntry> getTotalSubmissionsForExercise(@Param("startDate") ZonedDa
...
-            from AnswerPost answer left join answer.post post left join Channel channel ON channel.id = post.conversation.id
-            where channel.exercise.id = :#{#exerciseId} and answer.resolvesPost = true
+            SELECT COUNT(DISTINCT post.id)
+            FROM AnswerPost answer
+                LEFT JOIN answer.post post
+                LEFT JOIN TREAT (post.conversation AS Channel) channel
+            WHERE channel.exercise.id = :exerciseId
+                AND answer.resolvesPost IS TRUE
             """)
     long getNumberOfResolvedExercisePosts(@Param("exerciseId") Long exerciseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/StatisticsRepository.java:490]:
```suggestion
    long getNumberOfResolvedExercisePosts(@Param("exerciseId") long exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 14]
```diff
@@ -215,43 +256,73 @@ Page<User> searchAllByLoginOrNameInGroupsNotUserId(Pageable pageable, @Param("lo
...
+                JOIN Conversation conversation ON conversation.id = conversationParticipant.conversation.id
+            WHERE user.isDeleted IS FALSE
+                AND conversation.id = :conversationId
+                AND (
+                    :loginOrName = ''
+                    OR user.login LIKE :#{#loginOrName}%
+                    OR CONCAT_WS(' ', user.firstName, user.lastName) LIKE %:#{#loginOrName}%
+                )
             """)
     Page<User> searchAllByLoginOrNameInConversation(Pageable pageable, @Param("loginOrName") String loginOrName, @Param("conversationId") Long conversationId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:298]:
```suggestion
    Page<User> searchAllByLoginOrNameInConversation(Pageable pageable, @Param("loginOrName") String loginOrName, @Param("conversationId") long conversationId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 15]
```diff
@@ -260,14 +331,16 @@ Page<User> searchAllByLoginOrNameInConversationWithCourseGroups(Pageable pageabl
...
+                JOIN Conversation conversation ON conversation.id = conversationParticipant.conversation.id
+            WHERE user.isDeleted IS FALSE
+                AND conversation.id = :conversationId
+                AND (
+                    :loginOrName = ''
+                    OR user.login LIKE :#{#loginOrName}%
+                    OR CONCAT_WS(' ', user.firstName, user.lastName) LIKE %:#{#loginOrName}%
+                ) AND conversationParticipant.isModerator IS TRUE
             """)
     Page<User> searchChannelModeratorsByLoginOrNameInConversation(Pageable pageable, @Param("loginOrName") String loginOrName, @Param("conversationId") Long conversationId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:345]:
```suggestion
    Page<User> searchChannelModeratorsByLoginOrNameInConversation(Pageable pageable, @Param("loginOrName") String loginOrName, @Param("conversationId") long conversationId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 16]
```diff
@@ -349,31 +392,51 @@ default Page<UserDTO> searchAllUsersByLoginOrNameInGroupAndConvertToDTO(Pageable
...
+                    user.login LIKE :#{#loginOrName}%
+                    OR CONCAT_WS(' ', user.firstName, user.lastName) LIKE %:#{#loginOrName}%
+                )
+                AND (course.studentGroupName MEMBER OF user.groups
+                    OR course.teachingAssistantGroupName MEMBER OF user.groups
+                    OR course.editorGroupName MEMBER OF user.groups
+                    OR course.instructorGroupName MEMBER OF user.groups
+                )
             """)
     Page<User> searchAllByLoginOrNameInCourse(Pageable page, @Param("loginOrName") String loginOrName, @Param("courseId") Long courseId);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:425]:
```suggestion
    Page<User> searchAllByLoginOrNameInCourse(Pageable page, @Param("loginOrName") String loginOrName, @Param("courseId") long courseId);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 17]
```diff
@@ -349,31 +392,51 @@ default Page<UserDTO> searchAllUsersByLoginOrNameInGroupAndConvertToDTO(Pageable
...
 
     @Modifying
     @Transactional // ok because of modifying query
-    @Query("UPDATE User user SET user.lastNotificationRead = :#{#lastNotificationRead} WHERE user.id = :#{#userId}")
+    @Query("""
+            UPDATE User user
+            SET user.lastNotificationRead = :lastNotificationRead
+            WHERE user.id = :userId
+            """)
     void updateUserNotificationReadDate(@Param("userId") Long userId, @Param("lastNotificationRead") ZonedDateTime lastNotificationRead);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:440]:
```suggestion
    void updateUserNotificationReadDate(@Param("userId") long userId, @Param("lastNotificationRead") ZonedDateTime lastNotificationRead);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 18]
```diff
@@ -388,29 +451,48 @@ default Page<UserDTO> searchAllUsersByLoginOrNameInGroupAndConvertToDTO(Pageable
      */
     @Modifying
     @Transactional // ok because of modifying query
-    @Query("UPDATE User user SET user.hideNotificationsUntil = :#{#hideNotificationUntil} WHERE user.id = :#{#userId}")
+    @Query("""
+            UPDATE User user
+            SET user.hideNotificationsUntil = :hideNotificationUntil
+            WHERE user.id = :userId
+            """)
     void updateUserNotificationVisibility(@Param("userId") Long userId, @Param("hideNotificationUntil") ZonedDateTime hideNotificationUntil);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:459]:
```suggestion
    void updateUserNotificationVisibility(@Param("userId") long userId, @Param("hideNotificationUntil") ZonedDateTime hideNotificationUntil);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 19]
```diff
@@ -388,29 +451,48 @@ default Page<UserDTO> searchAllUsersByLoginOrNameInGroupAndConvertToDTO(Pageable
...
 
     @Modifying
     @Transactional // ok because of modifying query
-    @Query("UPDATE User user SET user.langKey = :#{#languageKey} WHERE user.id = :#{#userId}")
+    @Query("""
+            UPDATE User user
+            SET user.langKey = :languageKey
+            WHERE user.id = :userId
+            """)
     void updateUserLanguageKey(@Param("userId") Long userId, @Param("languageKey") String languageKey);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:468]:
```suggestion
    void updateUserLanguageKey(@Param("userId") long userId, @Param("languageKey") String languageKey);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 20]
```diff
@@ -388,29 +451,48 @@ default Page<UserDTO> searchAllUsersByLoginOrNameInGroupAndConvertToDTO(Pageable
...
 
     @Modifying
     @Transactional
-    @Query("UPDATE User user SET user.irisAccepted = :#{#acceptDatetime} WHERE user.id = :#{#userId}")
+    @Query("""
+            UPDATE User user
+            SET user.irisAccepted = :acceptDatetime
+            WHERE user.id = :userId
+            """)
     void updateIrisAcceptedToDate(@Param("userId") Long userId, @Param("acceptDatetime") ZonedDateTime acceptDatetime);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:477]:
```suggestion
    void updateIrisAcceptedToDate(@Param("userId") long userId, @Param("acceptDatetime") ZonedDateTime acceptDatetime);
```

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 21]
```diff
@@ -388,29 +451,48 @@ default Page<UserDTO> searchAllUsersByLoginOrNameInGroupAndConvertToDTO(Pageable
...
             FROM Team team
-            JOIN team.students st
-            WHERE st.isDeleted = false
-            AND team.exercise.course.id = :#{#courseId} AND team.shortName = :#{#teamShortName}
+                JOIN team.students st
+            WHERE st.isDeleted IS FALSE
+                AND team.exercise.course.id = :courseId
+                AND team.shortName = :teamShortName
             """)
     Set<User> findAllInTeam(@Param("courseId") Long courseId, @Param("teamShortName") String teamShortName);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:497]:
```suggestion
    Set<User> findAllInTeam(@Param("courseId") long courseId, @Param("teamShortName") String teamShortName);
```

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**